eras/alonzo/impl/testlib/Test/Cardano/Ledger/Alonzo/Binary/CostModelsSpec.hs

```diff
@@ -0,0 +1,156 @@
+                  cmRes <- expectRight cmDecoded
+                  cmRes `shouldSatisfy` invalidCm
+
+              -- in no era are we deserializing invalid costmodels within PParamsUpdate
+              expectDeserialiseFailure ppuDecoded Nothing
+  where
+    genInvalidCostModelsEnc lang = do
+      let validCount = costModelParamsCount lang
+      count <- choose (0, validCount - 1)
+      genCostModelsEncForLanguage lang count
+
+    invalidCm (CostModels valid errors _) =
+      null valid && (not . null) errors
+
+unknownCostModelProp ::
+  forall era.
+  AlonzoEraPParams era =>
+  Proxy era ->
+  Spec
+unknownCostModelProp pxy = do
+  prop "unknown CostModels deserialize correctly within PParamsUpdate starting with Conway" $
+    forAllShow genUnknownCostModelsEnc (showEnc pxy) $
+      \unknownCmEnc -> do
+        encodeAndCheckDecoded @era unknownCmEnc $
+          \cmDecoded ppuDecoded -> do
+            -- pre-Conway we are failing when deserializing unknown costmodels
+            if eraProtVerHigh @era < natVersion @9
+              then do
+                expectDeserialiseFailure cmDecoded (Just "CostModels")
+                expectDeserialiseFailure ppuDecoded Nothing
+              else do
+                -- post-Conway, we are collecting unknown CostModels
+                cmRes <- expectRight cmDecoded
+                cmRes `shouldSatisfy` unknownCm
+                ppuRes <- expectRight ppuDecoded
+                ppuRes `shouldSatisfy` \ppu -> (unknownCm <$> ppu ^. ppuCostModelsL) == SJust True
+  where
+    genUnknownCostModelsEnc = do
+      let firstUnknownLang = (fromEnum (maxBound @Language)) + 1
+      lang <- choose (fromIntegral firstUnknownLang, maxBound @Word8)
+      NonNegative count <- arbitrary
+      genCostModelsEnc lang count
+    unknownCm (CostModels valid errors unknown) =
+      null valid && null errors && (not . null) unknown
+
+encodeAndCheckDecoded ::
+  forall era.
+  AlonzoEraPParams era =>
+  Encoding ->
+  (Either DecoderError CostModels -> Either DecoderError (PParamsUpdate era) -> IO ()) ->
+  IO ()
+encodeAndCheckDecoded cmEnc check = do
+  let ppuEnc =
+        fromPlainEncoding
+          . toCBOR
+          . Map.singleton (18 :: Int)
+          . toPlainEncoding (eraProtVerHigh @era)
+          $ cmEnc
+  let cmBytes = serialize (eraProtVerHigh @era) cmEnc
+  let ppuBytes = serialize (eraProtVerHigh @era) ppuEnc
+  let ver = eraProtVerHigh @era
+  check
+    (decodeFull @CostModels ver cmBytes)
+    (decodeFull @(PParamsUpdate era) ver ppuBytes)
+
+genCostModelsEnc :: Word8 -> Int -> Gen Encoding
+genCostModelsEnc lang count = do
+  values <- vectorOf count (arbitrary :: Gen Int)
+  pure $ fromPlainEncoding . toCBOR $ Map.singleton lang values
+
+genCostModelsEncForLanguage :: Language -> Int -> Gen Encoding
+genCostModelsEncForLanguage lang =
+  genCostModelsEnc (fromIntegral (fromEnum lang))
+
+expectDeserialiseFailure :: Show t => Either DecoderError t -> Maybe Text -> IO ()
```

Comment:
Adding a call stack will inform you which usage of the function triggered the expectation failure
```suggestion
expectDeserialiseFailure :: 
  (HasCallStack, Show t) => Either DecoderError t -> Maybe Text -> IO ()
```